pkg/controller/sharding/scheduler.go

```diff
@@ -0,0 +1,215 @@
+			in.candidates[id] = map[string]bool{}
+		}
+		in.candidates[id][pod.Name] = healthy
+	}
+}
+
+func (in *dynamicDiscoveryScheduler) _unregisterPod(obj interface{}) {
+	if pod, ok := obj.(*corev1.Pod); ok {
+		id := pod.GetLabels()[LabelKubeVelaShardID]
+		klog.Infof("dynamicDiscoveryScheduler unregister pod %s/%s", pod.Namespace, pod.Name)
+		in.mu.Lock()
+		defer in.mu.Unlock()
+		if _, exist := in.candidates[id]; exist {
+			delete(in.candidates[id], pod.Name)
+			if len(in.candidates[id]) == 0 {
+				delete(in.candidates, id)
+			}
+		}
+	}
+}
+
+// resync the available shards
+func (in *dynamicDiscoveryScheduler) resync(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		default:
+			in.mu.Lock()
+			in.candidates = map[string]map[string]bool{}
+			in.mu.Unlock()
+			for _, obj := range in.store.List() {
+				in._registerPod(obj)
+			}
+			available := in.availableShards()
+			klog.Infof("dynamicDiscoveryScheduler resync finished, available shards: [%s]", strings.Join(available, ", "))
+		}
+		time.Sleep(in.resyncPeriod)
+	}
+}
+
+// Start run scheduler to watch pods and automatic register
+func (in *dynamicDiscoveryScheduler) Start(ctx context.Context) {
+	klog.Infof("dynamicDiscoveryScheduler staring, watching pods in %s", util.GetRuntimeNamespace())
+	cli := singleton.StaticClient.Get().CoreV1().RESTClient()
+	lw := cache.NewFilteredListWatchFromClient(cli, "pods", util.GetRuntimeNamespace(), func(options *metav1.ListOptions) {
+		ls := labels.NewSelector()
+		ls = ls.Add(*velaruntime.Must(labels.NewRequirement(LabelKubeVelaShardID, selection.Exists, nil)))
+		ls = ls.Add(*velaruntime.Must(labels.NewRequirement("app.kubernetes.io/name", selection.Equals, []string{in.name})))
+		options.LabelSelector = ls.String()
+	})
+	in.store, in.informer = cache.NewInformer(lw, &corev1.Pod{}, in.resyncPeriod, cache.ResourceEventHandlerFuncs{
+		AddFunc: in._registerPod,
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			if k8s.GetLabel(oldObj.(runtime.Object), LabelKubeVelaShardID) != k8s.GetLabel(newObj.(runtime.Object), LabelKubeVelaShardID) {
+				in._unregisterPod(oldObj)
+			}
+			in._registerPod(newObj)
+		},
+		DeleteFunc: in._unregisterPod,
+	})
+	stopCh := ctx.Done()
+	if stopCh == nil {
+		stopCh = make(chan struct{})
+	}
+	if in.resyncPeriod > 0 {
+		go in.resync(stopCh)
+	}
+	klog.Infof("dynamicDiscoveryScheduler started")
+	in.informer.Run(stopCh)
+}
+
+func (in *dynamicDiscoveryScheduler) availableShards() []string {
+	in.mu.RLock()
+	defer in.mu.RUnlock()
+	var available []string
+	for id, pods := range in.candidates {
+		if slices.Any(maps.Values(pods), func(x bool) bool { return x }) {
+			available = append(available, id)
+		}
+	}
+	return available
+}
+
+func (in *dynamicDiscoveryScheduler) schedule() (string, bool) {
+	available := in.availableShards()
+	if len(available) == 0 {
+		return "", false
+	}
+	// nolint
+	return available[rand.Intn(len(available))], true
```

Comment:
Advice uses Round Robin as the default schedule rule.

Reply:
Good idea.

Reply:
I don't prefer the schedule algorithm, but do we expose metrics of how many applications the controller is handling?  This is important.

Reply:
This is exposed by the standard metrics endpoint in Prometheus format. We need to add support for the Grafana to collect them.

Reply:
> Advice uses Round Robin as the default schedule rule.

Added to the dynamicDiscoveryScheduler. The staticDiscoveryScheduler will use random still.